### PR TITLE
TVP as an argument of a procedure does not require TypeName to be specified

### PR DESCRIPTION
**Pull request by tanyagupta17**
### Description

The type name of the specified table-valued parameter as an argument for a procedure should be optional. Currently, babelfish does not support this since we need explicit type to create the temp table.

We are fixing it by doing a look up of the typename of the TVP argument based on procedure name, namespace, db and role and setting the same.

Signed-off-by: Tanya Gupta [tanyagp@amazon.com](mailto:tanyagp@amazon.com)

### Issues Resolved

BABEL-5311

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** NA

* **Arbitrary inputs -** Yes

* **Negative test cases -** NA

* **Minor version upgrade tests -** Yes

* **Major version upgrade tests -** Yes

* **Performance tests -** NA

* **Tooling impact -** NA

* **Client tests -** Yes

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 1]
```diff
@@ -269,6 +288,104 @@ typedef TDSRequestData *TDSRequest;
 #define TVP_COLUMN_ORDERING_TOKEN		0x11
 #define TVP_END_TOKEN				0x00
 
+/* Find namespace oid of a procedure based on proc name. */
+static inline Oid
+get_proc_namespace_oid(char **proc_name, char *curr_db)
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: Why are we doing all of this in TDS?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 2]
```diff
@@ -4800,6 +4800,10 @@ _PG_init(void)
 		(*pltsql_protocol_plugin_ptr)->tsql_char_input = common_utility_plugin_ptr->tsql_bpchar_input;
 		(*pltsql_protocol_plugin_ptr)->get_cur_db_name = &get_cur_db_name;
 		(*pltsql_protocol_plugin_ptr)->get_physical_schema_name = &get_physical_schema_name;
+		(*pltsql_protocol_plugin_ptr)->get_authid_user_ext_schema_name = &get_authid_user_ext_schema_name;
+		(*pltsql_protocol_plugin_ptr)->fetch_func_input_arg_names = &fetch_func_input_arg_names;
+		(*pltsql_protocol_plugin_ptr)->split_object_name = &split_object_name;
+		(*pltsql_protocol_plugin_ptr)->construct_unique_hash = &construct_unique_hash;
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: This wont be required if we simply implement a direct API call in pltsql which can be exposed

File: test/dotnet/ExpectedOutput/TestTvp.out [hunk 3]
```diff
@@ -17,12 +17,12 @@
 #Q#drop type this_tvp_type_name_is_very_loooooooooooooooooooooooooooooooooooooooooong
 #Q#Create table tvp_table(a int, b smallint, c bigint, d tinyint, e bit, f char(10), g nchar(10), h varchar(10), i nvarchar(10), l varbinary(10), m binary(10), n date, o datetime, p money, q uniqueidentifier,r float, s real, t numeric(4,3), u decimal(5,3), v time(5), w datetime2(5))
 #Q#create type this_tvp_type_name_is_very_loooooooooooooooooooooooooooooooooooooooooong as table (a int, b smallint, c bigint, d tinyint, e bit, f char(10), g nchar(10), h varchar(10), i nvarchar(10), l varbinary(10), m binary(10), n date, o datetime, p money, q uniqueidentifier,r float, s real, t numeric(4,3), u decimal(5,3), v time(5), w datetime2(5))
-#Q#create procedure tvp_proc @this_tvp_parameter_name_is_very_loooooooooooooooooooooooooooooooooooooooooong as this_tvp_type_name_is_very_loooooooooooooooooooooooooooooooooooooooooong READONLY AS BEGIN insert into tvp_table select * from @this_tvp_parameter_name_is_very_loooooooooooooooooooooooooooooooooooooooooong END
```
Comment (KushaalShroff) [inline test/dotnet/ExpectedOutput/TestTvp.out]: Why this change?

File: test/dotnet/input/Datatypes/TestTvp.txt [hunk 4]
```diff
@@ -30,4 +31,63 @@ create schema this_schema_name_is_also_very_looooooooooooooooooooooooooooooooooo
...
 prepst#!#Select * from @this_tvp_parameter_name_is_very_loooooooooooooooooooooooooooooooooooooooooong#!#tvp|-|this_tvp_parameter_name_is_very_loooooooooooooooooooooooooooooooooooooooooong|-|this_schema_name_is_also_very_loooooooooooooooooooooooooooooooooooooooooong.this_tvp_type_name_is_very_loooooooooooooooooooooooooooooooooooooooooong|-|../../../utils/tvp-dotnet.csv
 drop type this_schema_name_is_also_very_loooooooooooooooooooooooooooooooooooooooooong.this_tvp_type_name_is_very_loooooooooooooooooooooooooooooooooooooooooong
-drop schema this_schema_name_is_also_very_loooooooooooooooooooooooooooooooooooooooooong
\ No newline at end of file
+drop schema this_schema_name_is_also_very_loooooooooooooooooooooooooooooooooooooooooong
+
+create table t(a int)
+create type t1 as table (a int)
+create procedure proc1 @t2 as t1 readonly as begin set nocount on; insert into t (a) select a from @t2 end
+storedproc#!#prep#!#proc1#!#tvp|-|t2|-| |-|../../../utils/tvp-test-dotnet.csv
```
Comment (KushaalShroff) [inline test/dotnet/input/Datatypes/TestTvp.txt:40]: This is the same tests which were present and does not test the current use case in mind, How are we testing the change you have made?

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 5]
```diff
@@ -15,14 +15,33 @@
...
+#include "utils/acl.h"
+#include "utils/fmgroids.h"
+#include "access/table.h"
+#include "access/genam.h"
+#include "utils/lsyscache.h"
+#include "access/xact.h"
+#include "utils/varlena.h"
+#include "catalog/pg_type.h"
+#include "parser/scansup.h"
+#include "parser/parser.h"
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: Too many unnecessary changes in tds extension

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 6]
```diff
@@ -536,21 +666,60 @@ SetColMetadataForTvp(ParameterToken temp, const StringInfo message, uint64_t *of
...
+				db_prefix = palloc0(BBF_NAMEDATALEND);
+				snprintf(db_prefix, BBF_NAMEDATALEND , "%s_", curr_db);
+
+				/* Remove db_prefix from begining of typnamespace to get the tvpTypeSchemaName*/
+				if (strncmp(typnamespace, db_prefix, strlen(db_prefix)) == 0) {
+					tvpschemaname = typnamespace + strlen(db_prefix);
+					oldContext = MemoryContextSwitchTo(TopMemoryContext);
+					temp->len += strlen(tvpschemaname);
+					temp->tvpInfo->tvpTypeSchemaName = pstrdup(tvpschemaname);
+					MemoryContextSwitchTo(oldContext);
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: Why are we doing this manually? There are already functions for this like get_logical_schema_name. Please check for internal utility functions since most of them should be available

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 7]
```diff
@@ -536,21 +666,60 @@ SetColMetadataForTvp(ParameterToken temp, const StringInfo message, uint64_t *of
...
+			/* check if user have right permission on object */
+			if (HeapTupleIsValid(tuple) && object_aclcheck(TypeRelationId, tvp_proargtype, user_id, ACL_USAGE) == ACLCHECK_OK)
+			{
+				Form_pg_type pg_type = (Form_pg_type) GETSTRUCT(tuple);
+				typename = NameStr(pg_type->typname);
+				typnamespace_oid = pg_type->typnamespace;
+				typnamespace = get_namespace_name(typnamespace_oid);
+
+				db_prefix = palloc0(BBF_NAMEDATALEND);
+				snprintf(db_prefix, BBF_NAMEDATALEND , "%s_", curr_db);
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: Not required, check earlier comment

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 8]
```diff
@@ -536,21 +666,60 @@ SetColMetadataForTvp(ParameterToken temp, const StringInfo message, uint64_t *of
...
-							"Table-valued parameter %d, to a parameterized string has no table type defined.",
-							temp->paramOrdinal + 1)));
+			if (!xactStarted)
+				StartTransactionCommand(); 
+			user_id = GetSessionUserId();
+			target_arg_name = temp->paramMeta.colName.data;
+			curr_db = pltsql_plugin_handler_ptr->get_cur_db_name();
+
+			/* Get procedure namespaceid. */
+			obj_schema_oid = get_proc_namespace_oid(&proc_name,curr_db);
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: space between ","

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 9]
```diff
@@ -269,6 +288,104 @@ typedef TDSRequestData *TDSRequest;
...
+get_proc_namespace_oid(char **proc_name, char *curr_db)
+{
+	char *logical_sch_name;
+	char *physical_sch_name;
+	char *db_name;
+	char *schema_name;
+	char	  **splited_object_name;
+	Oid obj_schema_oid = InvalidOid;
+
+	/* resolve the three part name */
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: Comment style is inconsistent. Please follow the appropriate guidelines.

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 10]
```diff
@@ -269,6 +288,104 @@ typedef TDSRequestData *TDSRequest;
...
+	schema_name = splited_object_name[2];
+	*proc_name = splited_object_name[3];
+
+	if (!strcmp(db_name, ""))
+		db_name = curr_db;
+
+	if (!strcmp(schema_name, ""))
+	{
+		/* Find the default schema for current user and get physical schema name */
+		const char *user = pltsql_plugin_handler_ptr->pltsql_get_user_for_database(db_name);
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: No need for const declaration. Cast to char*

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 11]
```diff
@@ -269,6 +288,104 @@ typedef TDSRequestData *TDSRequest;
...
+		schema_name = pltsql_plugin_handler_ptr->get_authid_user_ext_schema_name((const char *) db_name, user);
+	}
+
+	logical_sch_name = downcase_truncate_identifier(schema_name,strlen(schema_name), true);
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: Why downcase_truncate_identifier when we are reading from catalog?
This Should be in else

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 12]
```diff
@@ -269,6 +288,104 @@ typedef TDSRequestData *TDSRequest;
...
+static inline Oid
+tds_get_proargtypes_oid(char *proname, Oid pronamespace, Oid user_id, char *targeted_arg_name)
+{
+	HeapTuple    tuple;
+	CatCList   *catlist;
+	Oid matched_type = InvalidOid;
+	char *hash;
+	char new_target_name[NAMEDATALEN];
+
+	if (strlen(targeted_arg_name) >= NAMEDATALEN)
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: truncate identifier does this, we need to reuse the utility functions rather than implementing by self every time

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 13]
```diff
@@ -536,21 +666,60 @@ SetColMetadataForTvp(ParameterToken temp, const StringInfo message, uint64_t *of
...
 		{
-			/* Throw error if TabelType-Name is not provided */
-			ereport(ERROR,
-					(errcode(ERRCODE_PROTOCOL_VIOLATION),
-					 errmsg("The incoming tabular data stream (TDS) remote procedure call (RPC) protocol stream is incorrect. "
-							"Table-valued parameter %d, to a parameterized string has no table type defined.",
-							temp->paramOrdinal + 1)));
+			if (!xactStarted)
+				StartTransactionCommand(); 
+			user_id = GetSessionUserId();
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: This is incase of cross database, we might need to use GetUserId. Please check with harsh

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 14]
```diff
@@ -269,6 +288,104 @@ typedef TDSRequestData *TDSRequest;
...
+
+		if (procform->pronamespace == pronamespace &&
+			object_aclcheck(ProcedureRelationId, procform->oid, user_id, ACL_EXECUTE) == ACLCHECK_OK)
+		{
+			char **proargnames = pltsql_plugin_handler_ptr->fetch_func_input_arg_names(tuple);
+			Oid *proargtypes = procform->proargtypes.values;
+
+			for (int j = 0; j < procform->pronargs; j++)
+			{
+				if (strcmp(proargnames[j], targeted_arg_name) == 0)
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: Lets add tests with same name of args in the procedure, if its possible

File: contrib/babelfishpg_tds/src/backend/tds/tdsrpc.c [hunk 15]
```diff
@@ -1346,6 +1346,7 @@ ReadParameters(TDSRequestSP request, uint64_t offset, StringInfo message, int *p
 	TdsIoFunctionInfo tempFuncInfo;
 	uint16		paramOrdinal = 0;
 	int			retStatus;
+	char *proc_name;
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/backend/tds/tdsrpc.c]: Why does this need to declared in this scope?

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 16]
```diff
@@ -15,14 +15,33 @@
 #ifndef TDS_REQUEST_H
 #define TDS_REQUEST_H
 
+#define BBF_NAMEDATALEND (NAMEDATALEN + 2)
+#define MD5_HASH_LEN 32
+
 #include "postgres.h"
+#include "port.h"
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: Why port?

**Review by KushaalShroff — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 1 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: Moved to pltsql_utils

File: test/dotnet/ExpectedOutput/TestTvp.out (hunk 3 above)
Comment (tanyagupta17) [reply test/dotnet/ExpectedOutput/TestTvp.out]: Removed

File: test/dotnet/input/Datatypes/TestTvp.txt (hunk 4 above)
Comment (tanyagupta17) [reply test/dotnet/input/Datatypes/TestTvp.txt:40]: I am not passing the typeName "t1" explicitly unlike the tests above but rather an one space char which is trimmed in dotnet.

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 5 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: Removed. moved it to pltsql_utils

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 6 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: Modified. Using get_logical_schema_name now

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 7 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: Removed

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 8 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: Done

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 9 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: Done

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 10 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: To find the default schema name next we call`get_authid_user_ext_schema_name` and this requires `user` to be a const, that's why const declaration is needed for user

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 14 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: A procedure can't have two arguments with same name, variable names must be unique within a query batch or stored procedure.
Babelfish 
```
1> CREATE PROCEDURE proc21  @t2 t1 READONLY, @t2 INT AS BEGIN SET NOCOUNT ON; select 1 end
2> go
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
parameter name "@t2" used more than once
```
TSQL
```
1> CREATE PROCEDURE proc21  @t2 t1 READONLY, @t2 INT AS BEGIN SET NOCOUNT ON; select 1 end
2> go
Msg 134, Level 15, State 1, Server EC2AMAZ-EULFEOJ, Procedure proc21, Line 1
The variable name '@t2' has already been declared. Variable names must be unique within a query batch or stored procedure.
```

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 16 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: Removed

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 13 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: As currently we don't support cross-db for TVP proc, changed it to GetUserId

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 17]
```diff
@@ -2509,3 +2510,170 @@ get_owner_of_schema(const char *schema)
...
+	HeapTuple		tuple;
+	Oid 			typnamespace_oid = InvalidOid;
+	char 			*typnamespace;
+	char 			*curr_db;
+	MemoryContext 	oldContext;
+
+	if (!xactStarted)
+		StartTransactionCommand();
+	user_id = GetUserId();
+	curr_db = get_cur_db_name();
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: In case of cross-db execution,
1. When we reach here what is the current_user -> TSQL user or TSQL login?
2. What is the current database context? Asking if we are switching the database before executing this procedure.

Comment (tanyagupta17) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: We currently do not support Cross-db execution for procedure having TVP as an argument. In `TdsRecvTypeTable` here , we use the current db as the default type database. 
https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_4_X_DEV/contrib/babelfishpg_tds/src/backend/tds/tdstypeio.c#L2323
And we do not allow the user to explicitly send the type database either.
https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_4_X_DEV/contrib/babelfishpg_tds/src/include/tds_request.h#L518

File: contrib/babelfishpg_tds/src/backend/tds/tdsrpc.c [hunk 18]
```diff
@@ -1580,14 +1580,15 @@ ReadParameters(TDSRequestSP request, uint64_t offset, StringInfo message, int *p
 				}
 				break;
 			case TDS_TYPE_TABLE:
-				{
+				{	char *proc_name;
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/backend/tds/tdsrpc.c]: Why do we need proc_name?

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 19]
```diff
@@ -2509,3 +2510,137 @@ get_owner_of_schema(const char *schema)
...
+	char *db_name;
+	char *schema_name;
+	char	  **splited_object_name;
+	Oid obj_schema_oid = InvalidOid;
+
+	/* Resolve the three part name. */
+	splited_object_name = split_object_name(*proc_name);
+	db_name = splited_object_name[1];
+	schema_name = splited_object_name[2];
+	*proc_name = splited_object_name[3];
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Can any of them be NULL?

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 20]
```diff
@@ -2509,3 +2510,137 @@ get_owner_of_schema(const char *schema)
 
 	return result;
 }
+
+/* Find namespace oid of a procedure based on proc name. */
+static Oid
+get_proc_namespace_oid(char **proc_name, char *curr_db)
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Lets move this to catalog.c

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 21]
```diff
@@ -2509,3 +2510,137 @@ get_owner_of_schema(const char *schema)
...
+	physical_sch_name = get_physical_schema_name(db_name, schema_name);
+	/* Get namespace oid from physical schema name */
+	obj_schema_oid = get_namespace_oid(physical_sch_name, false);
+
+	return obj_schema_oid;
+
+}
+
+static Oid
+get_proargtypes_oid(char *proname, Oid pronamespace, Oid user_id, char *targeted_arg_name)
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c:2583]: comment summary at top, it should also give usage insights like we are using it for tvp to find the type id of the param

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 22]
```diff
@@ -2509,3 +2510,137 @@ get_owner_of_schema(const char *schema)
...
+	CatCList   *catlist;
+	Oid matched_type = InvalidOid;
+
+	targeted_arg_name = downcase_truncate_identifier(targeted_arg_name, strlen(targeted_arg_name), true);
+
+	/*
+	 * Search pg_proc for the procedure by name in the specified namespace and
+	 * return the argument type, if the targeted argument name matches
+	 */
+	catlist = SearchSysCacheList1(PROCNAMEARGSNSP, CStringGetDatum(proname));
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Can we directly filter with pronamespace since overloading is restricted?

File: contrib/babelfishpg_tds/src/include/tds_request.h [hunk 23]
```diff
@@ -498,16 +499,20 @@ SetTvpRowData(ParameterToken temp, const StringInfo message, uint64_t *offset)
...
+	uint16			isTvpNull;
+	char			*tempString;
+	int				i = 0;
+	char	   		*messageData = message->data;
+	StringInfo		tempStringInfo = palloc(sizeof(StringInfoData));
+	uint32_t		collation;
+
+	char			*tvp_type_name = NULL;
+	char 			*tvp_type_schema_name = NULL;
+	char 			*target_arg_name;
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tds/src/include/tds_request.h]: Dont really need any of these new variables in this scope

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 24]
```diff
@@ -2509,3 +2510,170 @@ get_owner_of_schema(const char *schema)
...
+	/* downcase identifier if needed */
+	if (pltsql_case_insensitive_identifiers)
+	{
+		db_name = downcase_identifier(db_name, strlen(db_name), false, false);
+		schema_name = downcase_identifier(schema_name, strlen(schema_name), false, false);
+		*proc_name = downcase_identifier(*proc_name, strlen(*proc_name), false, false);
+		for (int j = 0; j < 4; j++)
+			pfree(splited_object_name[j]);
+	}
+	else
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Seems like same code has been reused multiple places, possibilities of creating a utility?

File: contrib/babelfishpg_tds/src/backend/tds/tdsrpc.c (hunk 18 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/backend/tds/tdsrpc.c]: Removed

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 19 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: None of these could be NULL as per the definition`split_object_name` we fill unspecified parts with empty strings.
I have added a check that the incoming `proc_name` should not be null and then only we will call splited_object_name on it.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 20 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: Done

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 22 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: As per our offline discussion, this won't be possible, as PG allows overloading and to search by procedure name, and namespace we would also require arg list, which is not available in our case.

File: contrib/babelfishpg_tds/src/include/tds_request.h (hunk 23 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tds/src/include/tds_request.h]: Removed unnecessary arg and moved rest in else block to limit its scope

**Review by HarshLunagariya — APPROVED** [review]

**Review by KushaalShroff — APPROVED** [review]
